### PR TITLE
Add GetDockerHostIP

### PR DESCRIPTION
**Pull request by leeavital**
### What does this PR do?

GetDockerHostIP will loop through:
- configuration 
- procfs (if in host mode, left out of this PR)
- the ec2 metadata endpoint

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: pkg/util/docker/host_ip_provider.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+			log.Debugf("could not deduce host IP from source %s: %s", err)
+		} else {
+			return ips
+		}
+	}
+	return nil
+}
+
+func getHostIPFFromConfig() ([]string, error) {
+	hostIPs := config.Datadog.GetStringSlice("process_agent_config.host_ips")
```
Comment (p-lambert) [inline pkg/util/docker/host_ip_provider.go]: I think we should probably validate this input? `net.ParseIP` returns `nil` for invalid IPs.

File: pkg/util/ec2/ec2_test.go [hunk 2]
```diff
@@ -182,3 +182,23 @@ func TestGetInstanceIDMultipleVPC(t *testing.T) {
 	require.Error(t, err)
 	assert.Contains(t, err.Error(), "too many mac addresses returned")
 }
+
+func TestGetLocalIPv4(t *testing.T) {
```
Comment (p-lambert) [inline pkg/util/ec2/ec2_test.go]: 👍

File: pkg/util/docker/host_ip_provider.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	providers = append(providers, hostIPProvider{"ec2 metadata endpoint", ec2.GetLocalIPv4})
+	if isHostMode {
+		providers = append(providers, hostIPProvider{"/proc/net/route", DefaultHostIPs})
+	}
+
+	for _, attempt := range providers {
+		log.Debugf("attempting to get host ip from source: %s", attempt.name)
+		ips, err := attempt.provider()
+		if err != nil {
+			log.Debugf("could not deduce host IP from source %s: %s", err)
```
Comment (p-lambert) [inline pkg/util/docker/host_ip_provider.go]: I think you're missing an argument there (`attempt.name`)

Comment (p-lambert) [reply pkg/util/docker/host_ip_provider.go]: ah, you fixed it already. race condition 😛

File: pkg/util/docker/host_ip_provider.go (hunk 1 above)
Comment (leeavital) [reply pkg/util/docker/host_ip_provider.go]: Good call, done.

File: pkg/util/docker/host_ip_provider.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+		name     string
+		provider func() ([]string, error)
+	}
+
+	var isHostMode bool
+	if mode, err := GetAgentContainerNetworkMode(); err != nil && mode == "host" {
+		isHostMode = true
+	}
+
+	var providers []hostIPProvider
```
Comment (sunhay) [inline pkg/util/docker/host_ip_provider.go]: you can simplify this a bit by just doing 

```
type provider func() ([]string, error)

providers := map[string]provider{
		"config", getHostIPFFromConfig,
		"ec2 metadata endpoint", ec2.GetLocalIPv4,
}

if isHostMode {
		providers["/proc/net/route"] = DefaultHostIPs
}

for name, provider := range providers {
		ips, err := provider()
		...
}
```

Comment (sunhay) [reply pkg/util/docker/host_ip_provider.go]: Disregard - I just realized order matters here.

File: pkg/util/docker/host_ip_provider.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util/cache"
+	"github.com/DataDog/datadog-agent/pkg/util/ec2"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// GetDockerHostIP returns the IP address of the host. This is meant to be called
+// only when the agent is running in a dockerized environment.
+func GetDockerHostIP() []string {
```
Comment (leeavital) [inline pkg/util/docker/host_ip_provider.go]: I'll rename these to be plural (GetDockerHostIPs vs GetDockerHostIP)

**Review by p-lambert — APPROVED** [review]

**Review by hkaj — APPROVED** [review]

**Pull request merged**